### PR TITLE
[llvm_instcount] Leaderboard Submission: Random Search

### PR DESCRIPTION
**Pull request by ChrisCummins**
This adds a pure random policy that selects actions randomly until a                                                                                                                                                                    
fixed number of steps (the "patience" of the search) have been                                                                                                                                                                          
evaluated without an improvement to reward. The search stops after a                                                                                                                                                                    
predetermined amount of search time has elapsed.                                                                                                                                                                                        
                                                                                                                                                                                                                                        
The patience value was selected by running a random 200 searches on                                                                                                                                                                     
programs from the `blas-v0` dataset and selecting the value that                                                                                                                                                                        
produced the best average reward.  

## What is in this PR?

* An implementation of a random search agent, using the [eval_llvm_instcount_policy()](https://facebookresearch.github.io/CompilerGym/compiler_gym/leaderboard.html#compiler_gym.leaderboard.llvm_instcount.eval_llvm_instcount_policy) helper to run it on the codesize reduction task.
* A write-up of the approach, based on the [leaderboard submission template](https://github.com/facebookresearch/CompilerGym/blob/development/leaderboard/SUBMISSION_TEMPLATE.md).
* New leaderboard entries for results achieved using four different search time budgets:

|  | Walltime (mean) | Codesize Reduction (geomean) |
| --- | --- | --- |
| [Random search, p=1.25, t=10](results_p125_t10.csv) | 42.939s | 1.031× |
| [Random search, p=1.25, t=60](results_p125_t60.csv) | 91.215s | 1.045× |
| [Random search, p=1.25, t=3600](results_p125_t3600.csv) | 3,630.821s | 1.061× |
| [Random search, p=1.25, t=10800](results_p125_t10800.csv) | 10,512.356s | 1.062× |

### CODE REVIEW

File: leaderboard/llvm_codesize/random_search/README.md [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+**Did you modify the CompilerGym source code?**
+No.
+
+**What parameters does the approach have?**
+Search time *t*, patience *p*.
+
+**What range of values were considered for the above parameters?** Search time
+is fixed and is indicated by the leaderboard entry name, e.g. "Random search
+(t=30)" means a random search for a minimum 30 seconds. Eight values for
+patience were considered, *n/4*, *n/2*, *3n/4*, *n*, *5n/4*, *3n/2*, *7n/8*, and
```
Comment (hughleat) [inline leaderboard/llvm_codesize/random_search/README.md]: You do not explain what patience means.

Comment (hughleat) [reply leaderboard/llvm_codesize/random_search/README.md]: Ah, perhaps just add 'explained below' or similar?

File: leaderboard/llvm_codesize/random_search/README.md [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+for patience_ratio in 0.25 0.50 0.75 1.00 1.25 1.50 1.75 2.00 ; do
+    echo -e "\nEvaluating --patience_ratio=$patience_ratio"
+    logfile="blas_t${SEARCH_TIME}_${patience_ratio}.csv"
+    python random_search.py --max_benchmarks=20 --n=10 --dataset=blas-v0 \
+        --search_time="${SEARCH_TIME}" --patience_ratio="${PATIENCE_RATIO}" \
+        --logfile="$LOGFILE"
+    python -m compiler_gym.bin.validate --env=llvm-ic-v0 \
+        --reward_aggregation=geomean < "$LOGFILE" | tail -n4
+done
+```
```
Comment (hughleat) [inline leaderboard/llvm_codesize/random_search/README.md]: IMO, patience ratio is odd. Just straightforward patience makes more sense to me.
I can see that if you don't know the env, then you might need something generic, but not all envs will have fixed action spaces, so this won't help.

File: leaderboard/llvm_codesize/random_search/README.md [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+is fixed and is indicated by the leaderboard entry name, e.g. "Random search
+(t=30)" means a random search for a minimum 30 seconds. Eight values for
+patience were considered, *n/4*, *n/2*, *3n/4*, *n*, *5n/4*, *3n/2*, *7n/8*, and
+*2n*, where *n* is the size of the action space. The patience value was selected
+using the `blas-v0` dataset for validation, see appendix below.
+
+**Is the policy deterministic?**
+No.
+
+## Description
```
Comment (hughleat) [inline leaderboard/llvm_codesize/random_search/README.md]: What about Machine Description?

**Review by hughleat — APPROVED** [review]
LGTM. Few comments.

File: leaderboard/llvm_codesize/random_search/README.md (hunk 1 above)
Comment (ChrisCummins) [reply leaderboard/llvm_codesize/random_search/README.md]: Good idea thanks

File: leaderboard/llvm_codesize/random_search/README.md (hunk 3 above)
Comment (ChrisCummins) [reply leaderboard/llvm_codesize/random_search/README.md]: done

File: leaderboard/llvm_codesize/random_search/README.md (hunk 2 above)
Comment (ChrisCummins) [reply leaderboard/llvm_codesize/random_search/README.md]: Added extra description

File: leaderboard/llvm_codesize/random_search/README.md [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+
+The random search operates by selecting actions randomly until a fixed number of
+steps (the "patience" of the search) have been evaluated without an improvement
+to reward. The search stops after a predetermined amount of search time has
+elapsed.
+
+Pseudo-code for this search is:
+
+```c++
+float search_with_patience(CompilerEnv env, int patience, int search_time) {
```
Comment (JD-ETH) [inline leaderboard/llvm_codesize/random_search/README.md]: Why is the pseudo-code in C++? Is there a C++ interface too?

File: leaderboard/llvm_codesize/random_search/README.md [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+        --dataset="${VALIDATION_DATASET}" \
+        --search_time="${SEARCH_TIME}" \
+        --patience_ratio="${PATIENCE_RATIO}" \
+        --logfile="$LOGFILE"
+    python -m compiler_gym.bin.validate --env=llvm-ic-v0 \
+        --reward_aggregation=geomean < "$LOGFILE" | tail -n4
+done
+```
+
+The patience value that returned the best geomean reward can then be used as the
```
Comment (JD-ETH) [inline leaderboard/llvm_codesize/random_search/README.md]: Might be interesting to know how many rollout was done during the random search during the defined time period on a CPU cluster. But I guess it varies a lot depend on the environment in use.

**Review by JD-ETH — APPROVED** [review]
Looks good to me!

File: leaderboard/llvm_codesize/random_search/README.md [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+steps (the "patience" of the search) have been evaluated without an improvement
+to reward. The search stops after a predetermined amount of search time has
+elapsed.
+
+Pseudo-code for this search is:
+
+```c++
+float search_with_patience(CompilerEnv env, int patience, int search_time) {
+    float best_reward = -INFINITY;
+    int end_time = time() + search_time;
```
Comment (benoitsteiner) [inline leaderboard/llvm_codesize/random_search/README.md]: should we use a time_t instead of an integer to make the program more portable ?

**Review by benoitsteiner — APPROVED** [review]
LGTM

File: leaderboard/llvm_codesize/random_search/README.md (hunk 4 above)
Comment (ChrisCummins) [reply leaderboard/llvm_codesize/random_search/README.md]: No particular reason, it just feels more pseudo-y to me with types :)

We don't advertise a C++ interface, but it is entirely usable as a C++ library. Here's is an example random search. It looks pretty much the same as in python:

https://github.com/facebookresearch/CompilerGym/blob/development/examples/RandomSearch.cc#L107-L136

File: leaderboard/llvm_codesize/random_search/README.md (hunk 5 above)
Comment (ChrisCummins) [reply leaderboard/llvm_codesize/random_search/README.md]: Yep I agree that would be nice. We have the counts already, adding a logging printout to record them.

File: leaderboard/llvm_codesize/random_search/README.md (hunk 6 above)
Comment (ChrisCummins) [reply leaderboard/llvm_codesize/random_search/README.md]: This is just pseudo-code, but yeah I'll use `time_t` to make it clear that we're dealing with time units, thanks :)

**Pull request merged**